### ISSUE TITLE
Win32::Console problem

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4579]
Migrated from [rt.perl.org#7904](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7904) (status was 'rejected')

Searchable as RT7904$

**Comment by p5pRT** [comment]
### From andrew_harton@agilent.com


> Now again&#8203;: Can you produce a \*working\* \*subset\* of your current script
> that reproduces the error you had been mentioning?

After a lot of fiddling\, I have determined that the cause of the error was
something completely unrelated to what I first thought it was\.  Enclosed is
a small script that illustrates the problem;

\#\!/usr/bin/perl
use strict;
use warnings;
use Win32&#8203;::Console;

my $thing = Test\(\);
printf "Return is : $thing\\n";  \# This never gets printed\.\.\.

sub Test
\{
 my $OUT = new Win32&#8203;::Console\(STD\_OUTPUT\_HANDLE\);
 $OUT\->Cls\(\);
 $OUT\->WriteChar\("Press a key"\,1\,2\);
 \<STDIN>;
 return 2;
\}

Basically\, it seems that if a new console object is created in a subroutine\,
once the subroutine is exited\, it is no longer possible to print to the
screen\.  I have tried using $OUT\->Free\(\) at the end of the sub before
returning\, but that seems to make no difference\.
I've discovered that program execution continues just fine \(I can write to
an output filehandle\, for example\)\, but there's no screen output\.

Any pointers would be appreciated\.

Thanks\,

Andrew

**Comment by p5pRT** [comment]
### From @smpeters

> \[andrew\_harton@&#8203;agilent\.com \- Tue Nov 13 00&#8203;:33&#8203;:50 2001\]&#8203;:
> 
> 
> > Now again&#8203;: Can you produce a \*working\* \*subset\* of your current script
> > that reproduces the error you had been mentioning?
> 
> After a lot of fiddling\, I have determined that the cause of the error was
> something completely unrelated to what I first thought it was\. 
Enclosed is
> a small script that illustrates the problem;
> 
> \#\!/usr/bin/perl
> use strict;
> use warnings;
> use Win32&#8203;::Console;
> 
> my $thing = Test\(\);
> printf "Return is : $thing\\n";  \# This never gets printed\.\.\.
> 
> sub Test
> \{
>  my $OUT = new Win32&#8203;::Console\(STD\_OUTPUT\_HANDLE\);
>  $OUT\->Cls\(\);
>  $OUT\->WriteChar\("Press a key"\,1\,2\);
>  \<STDIN>;
>  return 2;
> \}
> 
> Basically\, it seems that if a new console object is created in a
subroutine\,
> once the subroutine is exited\, it is no longer possible to print to the
> screen\.  I have tried using $OUT\->Free\(\) at the end of the sub before
> returning\, but that seems to make no difference\.
> I've discovered that program execution continues just fine \(I can write to
> an output filehandle\, for example\)\, but there's no screen output\.
> 
> Any pointers would be appreciated\.
> 
> Thanks\,
> 
> Andrew
> 

Sorry\, but Win32&#8203;::Console is not part of the core Perl distribution\.  It
is part of the core ActiveState distribution\, so you can report the bug
to them at http&#8203;://bugs\.activestate\.com//ActivePerl/query\.cgi\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @demerphq

> Sorry\, but Win32&#8203;::Console is not part of the core Perl 
> distribution\.  It
> is part of the core ActiveState distribution\, so you can 
> report the bug
> to them at http&#8203;://bugs\.activestate\.com//ActivePerl/query\.cgi\.


Im just curious about this\. Is there a good reason why the core Win32
modules are NOT considered to be Core perl in general? Are there any other
OS'es where core perl doesn't deploy with what most users of that os would
consider to be standard modules?

yves

**Comment by p5pRT** [comment]
### From @rgs

Orton\, Yves wrote&#8203;:
> 
> Im just curious about this\. Is there a good reason why the core Win32
> modules are NOT considered to be Core perl in general? Are there any other
> OS'es where core perl doesn't deploy with what most users of that os would
> consider to be standard modules?

You should ask Sarathy\, he made most of those decisions regarding Win32
packaging \(see libwin32 on CPAN\.\)

The only core Win32 module is\, so far\, Win32 itself\.

Moreover I'm not very inclined to add modules in the core\.

\-\- 
\-\-I was just going to throw it away\, Mr Bloom said\.
Bantam Lyons raised his eyes suddenly and leered weakly\.
&nbsp;    \-\- Ulysses

**Comment by p5pRT** [comment]
### From @demerphq

Rafael Garcia\-Suarez said on 01 December 2004 18&#8203;:45
> Orton\, Yves wrote&#8203;:
> > 
> > Im just curious about this\. Is there a good reason why the core Win32
> > modules are NOT considered to be Core perl in general? Are there any
other
> > OS'es where core perl doesn't deploy with what most users of that os
would
> > consider to be standard modules?
> 
> You should ask Sarathy\, he made most of those decisions regarding Win32
> packaging \(see libwin32 on CPAN\.\)
> 
> The only core Win32 module is\, so far\, Win32 itself\.
> 
> Moreover I'm not very inclined to add modules in the core\.

Well\, 5\.6 is long gone\. And frankly so even is the company that Sarathy
worked for\, at least they are if you want to be picky about it\. "Activestate
a division of Sophos" is a different beast to "Activestate\, home of the 5\.6
Pumpking"\. I mean at what point will p5p stop relying on AS to maintain this
code? When they issue a formal stop support notice?

Consider that libwin32 hasn't been updated on CPAN since 08 Jul 2002
\(libwin32\-0\.191\)\. 

I am sympathetic to the point about not wanting to add a whole host of
modules to the core\, especially if they are specific to only one OS\, but IMO
it seems that libwin32 should at least be a p5p maintained package\.

Cheers\,

Yves

**Comment by p5pRT** [comment]
### From @rgs

Orton\, Yves wrote&#8203;:
> 
> Consider that libwin32 hasn't been updated on CPAN since 08 Jul 2002
> \(libwin32\-0\.191\)\. 
> 
> I am sympathetic to the point about not wanting to add a whole host of
> modules to the core\, especially if they are specific to only one OS\, but IMO
> it seems that libwin32 should at least be a p5p maintained package\.

Right\. Volunteers welcome\. One could even envision a special source tarball
for Win32 that bundles all libwin32 modules\. That's a nice idea\. Just find
somebody to do it\.

**Comment by p5pRT** [comment]
### From @jandubois

On Wed\, 01 Dec 2004\, Orton\, Yves wrote&#8203;:
> Consider that libwin32 hasn't been updated on CPAN since 08 Jul 2002
> \(libwin32\-0\.191\)\.

I have a copy of most/all patches that have been submitted to Sarathy
over the time and I plan to integrate them sometime early next year and
upload a new version to CPAN\.

Cheers\,
\-Jan

**Comment by p5pRT** [comment]
### From @demerphq

Jan Dubois wrote on 01 December 2004 19&#8203;:52
> On Wed\, 01 Dec 2004\, Orton\, Yves wrote&#8203;:
> > Consider that libwin32 hasn't been updated on CPAN since 08 Jul 2002
> > \(libwin32\-0\.191\)\.
> 
> I have a copy of most/all patches that have been submitted to Sarathy
> over the time and I plan to integrate them sometime early 
> next year and upload a new version to CPAN\.

Sounds great Jan\. Thanks a lot\. I look forward to seeing that\. 

Cheers\,
Yves

**Comment by p5pRT** [comment]
### From @ysth

On Wed\, Dec 01\, 2004 at 10&#8203;:52&#8203;:19AM \-0800\, Jan Dubois \<jand@&#8203;ActiveState\.com> wrote&#8203;:
> On Wed\, 01 Dec 2004\, Orton\, Yves wrote&#8203;:
> > Consider that libwin32 hasn't been updated on CPAN since 08 Jul 2002
> > \(libwin32\-0\.191\)\.
> 
> I have a copy of most/all patches that have been submitted to Sarathy
> over the time and I plan to integrate them sometime early next year and
> upload a new version to CPAN\.

Perhaps a new mailing list is in order?  If a perl\.org \[\*\]perl\-libwin32
list were started and you post what you have\, someone might take the
work off your shoulders\, and a beta release go out\.  Who actually has
the PAUSE rights to upload the module now?

\[\*\] this name is deliberately not perl\-win32\* since libwin32 is also
used by cygwin\.  Netware? wince?

**Comment by p5pRT** [comment]
### From @jandubois

On Wed\, 01 Dec 2004\, Yitzchak Scott\-Thoennes wrote&#8203;:
> Perhaps a new mailing list is in order? If a perl\.org \[\*\]perl\-libwin32
> list were started and you post what you have\, someone might take the
> work off your shoulders\, and a beta release go out\. Who actually has
> the PAUSE rights to upload the module now?
>
> \[\*\] this name is deliberately not perl\-win32\* since libwin32 is also
> used by cygwin\. Netware? wince?

Having a mailing list is a good idea\. I was thinking about \(ab\)using
perl\-win32\-porters for that\, as the real Win32 porting stuff is
happening on p5p anyways\. But I agree that a new mailinglist would be
better\. I don't really care were it is hosted; I can also get one setup
at ActiveState if that is easier\.

I believe Sarathy made me co\-maintainer of libwin32 on PAUSE\.  But since
libwin32 is a bundle of modules also maintained by other people\, with
separate PAUSE settings\, and I've never made an upload of libwin32 yet\,
we may see some problems the very first time I try to upload it\.

I'll be leaving to Europe on vacation after next week\, so I won't be able
to get anything done until January\.

Cheers\,
\-Jan

**Comment by p5pRT** [comment]
### From @ysth

Ask\, would you set up a perl\-libwin32@&#8203;perl\.org list?

On Wed\, Dec 01\, 2004 at 11&#8203;:56&#8203;:11AM \-0800\, Jan Dubois \<jand@&#8203;ActiveState\.com> wrote&#8203;:
> On Wed\, 01 Dec 2004\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > Perhaps a new mailing list is in order? If a perl\.org \[\*\]perl\-libwin32
> > list were started and you post what you have\, someone might take the
> > work off your shoulders\, and a beta release go out\. Who actually has
> > the PAUSE rights to upload the module now?
> >
> > \[\*\] this name is deliberately not perl\-win32\* since libwin32 is also
> > used by cygwin\. Netware? wince?
> 
> Having a mailing list is a good idea\. I was thinking about \(ab\)using
> perl\-win32\-porters for that\, as the real Win32 porting stuff is
> happening on p5p anyways\. But I agree that a new mailinglist would be
> better\. I don't really care were it is hosted; I can also get one setup
> at ActiveState if that is easier\.
> 
> I believe Sarathy made me co\-maintainer of libwin32 on PAUSE\.  But since
> libwin32 is a bundle of modules also maintained by other people\, with
> separate PAUSE settings\, and I've never made an upload of libwin32 yet\,
> we may see some problems the very first time I try to upload it\.
> 
> I'll be leaving to Europe on vacation after next week\, so I won't be able
> to get anything done until January\.
> 
> Cheers\,
> \-Jan
> 
>

**Comment by p5pRT** [comment]
### From @gsar

On Wed\, 01 Dec 2004 18&#8203;:24&#8203;:28 GMT\, "Orton\, Yves" wrote&#8203;:
>Rafael Garcia\-Suarez said on 01 December 2004 18&#8203;:45
>> Orton\, Yves wrote&#8203;:
>> > Im just curious about this\. Is there a good reason why the core Win32
>> > modules are NOT considered to be Core perl in general? Are there any
>other
>> > OS'es where core perl doesn't deploy with what most users of that os
>would
>> > consider to be standard modules?
>> 
>> You should ask Sarathy\, he made most of those decisions regarding Win32
>> packaging \(see libwin32 on CPAN\.\)

I suggest searching the archives for previous traffic on this\.

>> The only core Win32 module is\, so far\, Win32 itself\.
>> 
>> Moreover I'm not very inclined to add modules in the core\.
>
>Well\, 5\.6 is long gone\. And frankly so even is the company that Sarathy
>worked for\, at least they are if you want to be picky about it\. "Activestate
>a division of Sophos" is a different beast to "Activestate\, home of the 5\.6
>Pumpking"\.

There is probably some truth to this\, but I will have you know that
I still have an ActiveState\.com address and am technically still
employed by ActiveState\. :\)

It has been a year or more since I have passed on most my Perl
related maintainership responsibilities to Jan Dubois and
Gisle Aas\.  I trust either of them to do justice to everything
that deserves it\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From DavidA@ActiveState.com

 > Well\, 5\.6 is long gone\. And frankly so even is the company that
 > Sarathy worked for\, at least they are if you want to be picky about
 > it\. "Activestate a division of Sophos" is a different beast to
 > "Activestate\, home of the 5\.6 Pumpking"\. I mean at what point will p5p
 > stop relying on AS to maintain this code? When they issue a formal
 > stop support notice?

Hi p5p'ers \-\-

I couldn't let the above go by without comment\.

In brief \-\- things change\, but we're still here\, and still 100% 
committed to Perl\.

It's true that GSar doesn't spend much time on core Perl anymore\.  I'm 
sure none of you would begrudge him the fact that he led the development 
of a Perl\-based tool that grew into a hugely successful enterprise 
solution\, and feels a responsibility to help it grow and succeed\.  It is 
also true that gsar's job has made it hard for him to spend as much time 
on updating his CPAN modules\.

ActiveState sees its key role in the Perl space as helping in those 
areas where the open source community won't naturally address\.  A long 
time ago\, this focused on making Perl work well on Windows\.  Now\, with a 
strong and active Perl on Windows community\, we see our strength 
focusing more on packaging and distribution\, and\, of course\, tools\. This 
is why we still maintain & distribute ActivePerl\, run the PPM service\, 
which incorporate CPAN updates as frequently as we can make it\, and work 
with platform vendors like HP to ensure strong support for Perl across 
the operating system landscape\.

Some of our other ongoing Perl\-related efforts which you might not know 
about include&#8203;:

&nbsp;   \* Jan Dubois has taken over maintenance of the win32 modules from 
Sarathy\, and is scheduled to upgrading them in CPAN and in ActivePerl
&nbsp;   \* Jan is also incorporating the PerlEx technology into ActivePerl

I know it's not as important to the p5p crowd\, but I think the broader 
Perl community appreciates some of the other tool work we've done 
recently\, such as the brand\-spanking new Perl Dev Kit 6\.0\.  As soon as 
we resolve some technical issues\, we're planning on publishing on CPAN 
the Perl\-Tk module Gisle built as part of that release \(I think the Perl 
Dev Kit includes some of the nicest\-looking Tk\-based Perl apps ever 
built\, but I'm biased\)\.  We're also adding Perl autocomplete & calltips 
to the next minor release of Komodo\.

There are other projects ongoing that we can't talk about yet because of 
contractual constraints but that we believe will have significant value 
for Perl users in the long term\.

I hope this clarifies where ActiveState stands w\.r\.t\. Perl these days\. 
If you have any questions or concerns\, feel free to email me or Jan 
Dubois\, who'se in charge of our Perl initiatives\.

\-\- David Ascher
&nbsp;    Managing Director & Chief Technologist
&nbsp;    ActiveState \-\- Dynamic Tools for Dynamic Languages

PS&#8203;: I'm not subscribed to p5p so remember to cc&#8203;: me if you want me to 
read a reply =\)\.

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Dec 01\, 2004 at 05&#8203;:29&#8203;:17PM \-0800\, David Ascher wrote&#8203;:

> In brief \-\- things change\, but we're still here\, and still 100% 
> committed to Perl\.

They're also advertising jobs on http&#8203;://jobs\.perl\.org/
which isn't the sign of a company downsizing to nothing\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @demerphq

Gurusamy Sarathy wrote on 02 December 2004 01&#8203;:02
> On Wed\, 01 Dec 2004 18&#8203;:24&#8203;:28 GMT\, "Orton\, Yves" wrote&#8203;:
> >Rafael Garcia\-Suarez said on 01 December 2004 18&#8203;:45

snip

> >> Moreover I'm not very inclined to add modules in the core\.
> >
> >Well\, 5\.6 is long gone\. And frankly so even is the company  that Sarathy
> >worked for\, at least they are if you want to be picky about it\.
"Activestate
> >a division of Sophos" is a different beast to "Activestate\, home of the
5\.6
> >Pumpking"\.
> 
> There is probably some truth to this\, but I will have you know that
> I still have an ActiveState\.com address and am technically still
> employed by ActiveState\. :\)
> 
> It has been a year or more since I have passed on most my Perl
> related maintainership responsibilities to Jan Dubois and
> Gisle Aas\.  I trust either of them to do justice to everything
> that deserves it\.

Just to be clear here\, that wasn't intended at all to be a slight on
Activestate or you or the others working there\. On the contrary Im very
appreciative of ActiveState\.

Its just always struck me as strange that critical modules for one of Perls
well deployed to OS'es are under the control of a corporate entity and not
the P5P community in general\. Obviously a coporate concern can only allocate
so much time to support something like Perl and it would seem to be in the
interests of us all to position things so as to spread the load\. 

Cheers\,
Yves

**Comment by p5pRT** [comment]
### From @gsar

On Thu\, 02 Dec 2004 15&#8203;:13&#8203;:50 GMT\, "Orton\, Yves" wrote&#8203;:
>Its just always struck me as strange that critical modules for one of Perls
>well deployed to OS'es are under the control of a corporate entity and not
>the P5P community in general\.

I might agree with you if it were some nameless\, faceless Corporate
Entity\, but knowing what I know about ActiveState\, this sounds too
much like FUD to me\.

ActiveState the corporate entity is made of individuals too\.  Many
of these individuals have always been part of the P5P community\, and
get to exercise their free will over their community contributions\.
There are no sinister motives involved\.

>Obviously a coporate concern can only allocate
>so much time to support something like Perl and it would seem to be in the
>interests of us all to position things so as to spread the load\. 

Yes\, I'm sure Jan would welcome offers to help maintain libwin32\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From @demerphq

> I might agree with you if it were some nameless\, faceless Corporate
> Entity\, but knowing what I know about ActiveState\, this sounds too
> much like FUD to me\.

Argh\, obviously I've made things worse\. I was not trying to FUD here\.
Really\. Sorry it seems that way\.
 
> ActiveState the corporate entity is made of individuals too\.  Many
> of these individuals have always been part of the P5P community\, and
> get to exercise their free will over their community contributions\.
> There are no sinister motives involved\.
> 
> >Obviously a coporate concern can only allocate
> >so much time to support something like Perl and it would seem to be in
the
> >interests of us all to position things so as to spread the load\. 
> 
> Yes\, I'm sure Jan would welcome offers to help maintain libwin32\.

Im not sure what good my help would be\, but I'd be glad to try\. If theres
something I can do then Jan should send me a mail\. :\-\)

Cheers\,
Yves

**Comment by p5pRT** [comment]
### From @gsar

On Thu\, 02 Dec 2004 16&#8203;:45&#8203;:10 GMT\, "Orton\, Yves" wrote&#8203;:
>> Yes\, I'm sure Jan would welcome offers to help maintain libwin32\.
>
>Im not sure what good my help would be\, but I'd be glad to try\. If theres
>something I can do then Jan should send me a mail\. :\-\)

Thanks\.  I know Jan has a list of things to be done\, but another
good first step might be to follow up on the bugs in RT&#8203;:

&nbsp;    http&#8203;://rt\.cpan\.org/NoAuth/Bugs\.html?Dist=libwin32


Sarathy
gsar@&#8203;ActiveState\.com
